src/bgw/scheduler.c

```diff
@@ -241,13 +253,24 @@ scheduled_bgw_job_transition_state_to(ScheduledBgwJob *sjob, JobState new_state)
 			Assert(!sjob->reserved_worker);
 
 			StartTransactionCommand();
+			BgwJobLockResult res = ts_bgw_job_get_share_lock(sjob->job.fd.id, CurrentMemoryContext);
 
-			if (!ts_bgw_job_get_share_lock(sjob->job.fd.id, CurrentMemoryContext))
+			if (res != JOB_FOUND_AND_LOCKED)
 			{
-				elog(WARNING,
-					 "scheduler detected that job %d was deleted when starting job",
-					 sjob->job.fd.id);
-				scheduled_bgw_job_transition_state_to(sjob, JOB_STATE_SCHEDULED);
+				if (res == JOB_NOT_FOUND)
+				{
+					elog(WARNING,
+						 "scheduler detected that job %d was deleted when starting job",
+						 sjob->job.fd.id);
+					ts_bgw_job_cache_invalidate_callback();
```

Comment:
Since the pattern now is to invalidate the cache if the lock function returns `JOB_NOT_FOUND`, maybe the invalidation should move into the locking function where it returns that value?